Firestore/Source/Remote/FSTDatastore.mm

```diff
@@ -356,6 +359,24 @@ + (void)prepareHeadersForRPC:(GRPCCall *)rpc
       [FSTDatastore googleCloudResourcePrefixForDatabaseID:databaseID];
 }
 
++ (std::unique_ptr<grpc::ClientContext>)
+createGrpcClientContextWithDatabaseID:(const DatabaseId *)databaseID
+                                token:(const absl::string_view)token {
+  auto context = absl::make_unique<grpc::ClientContext>();
+
+  if (token.data()) {
+    context->set_credentials(grpc::AccessTokenCredentials(token.data()));
+  }
+
+  context->AddMetadata(util::MakeString(kXGoogAPIClientHeader),
+                      util::MakeString([FSTDatastore googAPIClientHeaderValue]));
+  // This header is used to improve routing and project isolation by the backend.
+  context->AddMetadata(
+      util::MakeString(kGoogleCloudResourcePrefix),
+      util::MakeString([FSTDatastore googleCloudResourcePrefixForDatabaseID:databaseID]));
```

Comment:
If you port these header value factories to C++ it's likely that you can remove this code from the Objective-C sources altogether.